core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -96,7 +97,12 @@ object RequestChannel extends Logging {
       else
         null
 
-    buffer = null
+    //most request types are parsed entirely into objects at this point. for those we can release the underlying buffer.
+    //some (like produce, or any time the schema contains fields of types BYTES or NULLABLE_BYTES) retain a reference
+    //to the buffer. for those requests we cannot release the buffer early, but only when request processing is done.
+    if (requestId != ApiKeys.PRODUCE.id && requestId != ApiKeys.JOIN_GROUP.id && requestId != ApiKeys.SYNC_GROUP.id) {
```

Comment:
@ewencp - i've implemented a simple schema visitor and used that to find the relevant API keys for this dynamically. please see the revised code.